feat: dkim registry (#1084)
ethereum/ERCs: ERCS/erc-7969.md

---
eip: 7969
title: DKIM Registry Interface

> description: An interface for registering and validating DKIM public key hashes onchain.

Reviewer: You can probably omit "Interface" from your title, since they're pretty common in ERCs, and use the space to add more detail about the proposal. Maybe something like "DKIM Registry for ..." and add the main use case?

The description needs a bit more work. You should probably explain what DKIM is, or how putting it on chain is beneficial, or really anything more than just repeating your title but with more words.

> description: A standard interface for registering and validating DKIM public key hashes onchain.

Author: Ok, got it

author: Mike Fu (@fumeng00mike), Matthew Yu (@0xknon), Ernesto García (@ernestognw)
discussions-to: https://ethereum-magicians.org/t/new-erc-dkim-registry-interface/24530
status: Draft
type: Standards Track
category: ERC
created: 2025-06-11
---

## Abstract

> This EIP proposes a standard interface for registering and validating DKIM (DomainKeys Identified Mail) public key hashes on the Ethereum blockchain. The interface allows domain owners to register their DKIM public key hashes and enables third parties to verify the validity of these hashes.

Reviewer: Not bad, but maybe a tad too brief. You can go into a bit more technical detail (while keeping it an overview). How do DKIM clients get the information, for example?

## Motivation

> With the growing adoption of Account Abstraction [ERC-4337](https://eips.ethereum.org/EIPS/eip-4337) and the emergence of ZK Email Technology, there is a need for a standardized way to verify email ownership on-chain. This EIP provides a crucial building block for these technologies by enabling the verification of DKIM signatures through on-chain registries.

Reviewer 2: ```suggestion
With the growing adoption of Account Abstraction [ERC-4337] and the emergence of ZK Email Technology, there is a need for a standardized way to verify email ownership on-chain. This EIP provides a crucial building block for these technologies by enabling the verification of DKIM signatures through on-chain registries.

[ERC-4337]: ./eip-4337.md 
```

This standard enables several important use cases:

1. **Account Abstraction**: When combined with zkEmail, this registry enables email-based account abstraction. Users can prove ownership of their email address through DKIM signatures, allowing them to:
   - Create and manage smart contract wallets
   - Sign transactions using their email credentials
   - Implement social recovery mechanisms
2. **Account Recovery**: The registry facilitates secure account recovery mechanisms:
   - Users can recover their wallet access by on-chain proving email ownership
   - The process is trustless and secure due to the cryptographic nature of DKIM

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

### Interface

```solidity

> // SPDX-License-Identifier: MIT

Reviewer: Code within the document itself must be CC0-1.0 licensed. We do permit MIT code in the `assets/` directory and you can link to it there.

pragma solidity ^0.8.25;
/// @title ERC-XXX DKIM Registry Interface
/// @dev See https://eips.ethereum.org/EIPS/eip-xxx
/// Note: the ERC-165 identifier for this interface is 0xdee3d600.
interface IDKIMRegistry {
    event KeyHashRegistered(bytes32 domainHash, bytes32 keyHash);
    event KeyHashRevoked(bytes32 domainHash);
    function isKeyHashValid(
        bytes32 domainHash,
        bytes32 keyHash
    ) external view returns (bool);
}
```

### Domain Hash

The `domainHash` parameter MUST be the keccak256 hash of the lowercase domain name or subdomain name. For example:

- For the domain "example.com":

```solidity
domainHash = keccak256(bytes("example.com"))
```

- For the subdomain "mail.example.com":

```solidity
domainHash = keccak256(bytes("mail.example.com"))
```

The registry MUST treat each domain and subdomain as a distinct entity. This means that:

1. A key hash registered for "example.com" does not automatically apply to its subdomains
2. Each subdomain can have its own independent DKIM key hash registration
3. The full domain name (including subdomain if present) must be hashed as a single string

### Key Hash

> The `keyHash` parameter MUST be the keccak256 hash of the DKIM public key. The public key should be in the standard DKIM format as specified in RFC 6376.

Reviewer: ```suggestion
The `keyHash` parameter MUST be the keccak256 hash of the DKIM public key. The public key should be in the standard DKIM format as specified in [RFC 6376](https://www.rfc-editor.org/rfc/rfc6376).
```

```solidity
function isKeyHashValid(
    bytes32 domainHash,
    bytes32 keyHash
) external view returns (bool)
```

## Rationale

The interface is designed to be simple and focused on the core functionality of DKIM public key hash registration and validation. The use of keccak256 hashing for both domain names and public keys ensures consistent and secure handling of the data.

The events allow for efficient tracking of key registrations and revocations, which is important for maintaining the integrity of the registry.

## Backwards Compatibility

This EIP introduces a new interface and does not affect existing contracts or standards.

## Security Considerations

1. Domain owners must ensure they have control over their private keys and domain names.
2. The registry implementation should include proper access control mechanisms to prevent unauthorized registrations.
3. The registry should implement a mechanism to handle key rotation and revocation.
4. Implementations should consider rate limiting to prevent spam registrations.

## Reference Implementation

```solidity

> // SPDX-License-Identifier: MIT

Reviewer: Same CC0-1.0 comment.

Author: Thank you for your comment. I directly removed the license as other ERCs usually don't have it as well.

pragma solidity ^0.8.0;
import "@openzeppelin/contracts/access/Ownable.sol";
import "./interfaces/IDKIMRegistry.sol";
contract DKIMRegistry is IDKIMRegistry, Ownable {

>     constructor(address _owner) Ownable(_owner) { }
>     // Mapping from hashed domain name to DKIM public key hash to enabled
>     mapping(bytes32 => mapping(bytes32 => bool)) private _keyHashes;
>     /**
>      * @notice Checks if a DKIM key hash is valid for a given domain
>      * @param domainHash The hash of the domain name
>      * @param keyHash The hash of the DKIM public key
>      * @return bool True if the key hash is valid for the domain, false otherwise
>      */
>     function isKeyHashValid(
>         bytes32 domainHash,
>         bytes32 keyHash
>     ) public view returns (bool) {
>         return _keyHashes[domainHash][keyHash];
>     }
>     /**
>      * @notice Sets a DKIM key hash for a domain
>      * @param domainHash The hash of the domain name
>      * @param keyHash The hash of the DKIM public key to register
>      * @dev Only callable by the contract owner
>      * @dev Cannot set zero hash as a valid key hash
>      */
>     function setKeyHash(
>         bytes32 domainHash,
>         bytes32 keyHash
>     ) public onlyOwner {
>         require(keyHash != bytes32(0), "cannot set zero hash");
>         _keyHashes[domainHash][keyHash] = true;
>         emit KeyHashRegistered(domainHash, keyHash);
>     }
>     /**
>      * @notice Sets multiple DKIM key hashes for a domain in a single transaction
>      * @param domainHash The hash of the domain name
>      * @param keyHashes Array of DKIM public key hashes to register
>      * @dev Only callable by the contract owner
>      * @dev Array must not be empty
>      * @dev Each key hash must not be zero
>      */
>     function setKeyHashes(
>         bytes32 domainHash,
>         bytes32[] memory keyHashes
>     ) public onlyOwner {
>         require(keyHashes.length > 0, "empty array");
>         for (uint256 i = 0; i < keyHashes.length; i++) {
>             setKeyHash(domainHash, keyHashes[i]);
>         }
>     }
>     /**
>      * @notice Revokes a DKIM key hash for a domain
>      * @param domainHash The hash of the domain name
>      * @param keyHash The hash of the DKIM public key to revoke
>      * @dev Only callable by the contract owner
>      * @dev Sets the key hash mapping to false, effectively revoking it
>      */
>     function revokeKeyHash(bytes32 domainHash, bytes32 keyHash) public onlyOwner {
>         delete _keyHashes[domainHash][keyHash];
>         emit KeyHashRevoked(domainHash);
>     }
> }
> ```

Reviewer 2: We need to switch the order of these blocks to make the CI

```suggestion
## Reference Implementation

\`\`\`solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;
import "@openzeppelin/contracts/access/Ownable.sol";
import "./interfaces/IDKIMRegistry.sol";
contract DKIMRegistry is IDKIMRegistry, Ownable {
    constructor(address _owner) Ownable(_owner) { }
    // Mapping from hashed domain name to DKIM public key hash to enabled
    mapping(bytes32 => mapping(bytes32 => bool)) private _keyHashes;
    /**
     * @notice Checks if a DKIM key hash is valid for a given domain
     * @param domainHash The hash of the domain name
     * @param keyHash The hash of the DKIM public key
     * @return bool True if the key hash is valid for the domain, false otherwise
     */
    function isKeyHashValid(
        bytes32 domainHash,
        bytes32 keyHash
    ) public view returns (bool) {
        return _keyHashes[domainHash][keyHash];
    }
    /**
     * @notice Sets a DKIM key hash for a domain
     * @param domainHash The hash of the domain name
     * @param keyHash The hash of the DKIM public key to register
     * @dev Only callable by the contract owner
     * @dev Cannot set zero hash as a valid key hash
     */
    function setKeyHash(
        bytes32 domainHash,
        bytes32 keyHash
    ) public onlyOwner {
        require(keyHash != bytes32(0), "cannot set zero hash");
        _keyHashes[domainHash][keyHash] = true;
        emit KeyHashRegistered(domainHash, keyHash);
    }
    /**
     * @notice Sets multiple DKIM key hashes for a domain in a single transaction
     * @param domainHash The hash of the domain name
     * @param keyHashes Array of DKIM public key hashes to register
     * @dev Only callable by the contract owner
     * @dev Array must not be empty
     * @dev Each key hash must not be zero
     */
    function setKeyHashes(
        bytes32 domainHash,
        bytes32[] memory keyHashes
    ) public onlyOwner {
        require(keyHashes.length > 0, "empty array");
        for (uint256 i = 0; i < keyHashes.length; i++) {
            setKeyHash(domainHash, keyHashes[i]);
        }
    }
    /**
     * @notice Revokes a DKIM key hash for a domain
     * @param domainHash The hash of the domain name
     * @param keyHash The hash of the DKIM public key to revoke
     * @dev Only callable by the contract owner
     * @dev Sets the key hash mapping to false, effectively revoking it
     */
    function revokeKeyHash(bytes32 domainHash, bytes32 keyHash) public onlyOwner {
        delete _keyHashes[domainHash][keyHash];
        emit KeyHashRevoked(domainHash);
    }
}
\`\`\`

## Security Considerations

1. Domain owners must ensure they have control over their private keys and domain names.
2. The registry implementation should include proper access control mechanisms to prevent unauthorized registrations.
3. The registry should implement a mechanism to handle key rotation and revocation.
4. Implementations should consider rate limiting to prevent spam registrations.
```

## Copyright

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 2: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

### Interface

```solidity
pragma solidity ^0.8.25;
/// @title ERC-XXX DKIM Registry Interface
/// @dev See https://eips.ethereum.org/EIPS/eip-xxx
/// Note: the ERC-165 identifier for this interface is 0xdee3d600.
interface IDKIMRegistry {
    event KeyHashRegistered(bytes32 domainHash, bytes32 keyHash);
    event KeyHashRevoked(bytes32 domainHash);
    function isKeyHashValid(
        bytes32 domainHash,
        bytes32 keyHash
    ) external view returns (bool);
}
```

### Domain Hash

The `domainHash` parameter MUST be the keccak256 hash of the lowercase domain name or subdomain name. For example:

- For the domain "example.com":

```solidity
domainHash = keccak256(bytes("example.com"))
```

- For the subdomain "mail.example.com":

```solidity
domainHash = keccak256(bytes("mail.example.com"))
```

The registry MUST treat each domain and subdomain as a distinct entity. This means that:

1. A key hash registered for "example.com" does not automatically apply to its subdomains
2. Each subdomain can have its own independent DKIM key hash registration
3. The full domain name (including subdomain if present) must be hashed as a single string

### Key Hash

> The `keyHash` parameter MUST be the keccak256 hash of the DKIM public key. The public key should be in the standard DKIM format as specified in [RFC 6376](https://www.rfc-editor.org/rfc/rfc6376).

Reviewer 2: Addressing [@zkfriendly feedback](https://ethereum-magicians.org/t/erc-7969-dkim-registry-interface/24530/7)

```suggestion
The `keyHash` parameter MUST be a cryptographic hash of the DKIM public key. The public key should be in the standard DKIM format as specified in [RFC 6376](https://www.rfc-editor.org/rfc/rfc6376).

Implementations MAY choose any cryptographically secure hash function for computing the key hash. Common choices include keccak256, Poseidon (for zk-friendly applications) or other hash functions.
```

### DKIM Public Key Specification

The DKIM public key MUST follow the format specified in RFC 6376. Here are the key requirements:

1. **Key Format**:
   - The public key MUST be in the format specified in the DKIM DNS record (p=PUBLIC_KEY)
   - The key MUST be base64 encoded
   - The key MUST NOT include the PEM headers or any other formatting
   - The key MUST be the raw public key data as specified in the DKIM DNS record
2. **Key Requirements**:
   - The key MUST be a valid RSA public key
   - The key MUST be in the format as published in the domain's DNS TXT record
   - The key MUST be the exact value from the p= parameter in the DKIM DNS record
   - The key MUST NOT include any whitespace or line breaks
3. **Key Registration Process**:
   1. Obtain the DKIM public key from the domain's DNS TXT record
   2. Extract the value from the p= parameter

>    3. Calculate the keccak256 hash of the raw public key

Reviewer 2: Accordingly

```suggestion
   3. Calculate the hash of the raw public key using the implementation's chosen hash function
```

   4. Register the hash in the DKIM registry
4. **Key Validation**:
   - The registry MUST verify that the provided key hash corresponds to a valid DKIM public key
   - The key MUST be in the correct format as specified in RFC 6376
   - The key MUST be the exact value as published in the domain's DNS record

> 5. **Example**:

Reviewer 2: ```suggestion
5. **Example with keccak256**:
```
